File: src/main/java/jp/tsubakicraft/mongocrud/controller/CarController.java
package jp.tsubakicraft.mongocrud.controller;

import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import jp.tsubakicraft.mongocrud.model.Car;
import jp.tsubakicraft.mongocrud.model.Model;
import jp.tsubakicraft.mongocrud.service.CarRepository;
import jp.tsubakicraft.mongocrud.service.ModelRepository;

@RestController
public class CarController {

	@Autowired
	private CarRepository repo;

	@Autowired
	private ModelRepository modelRepo;

	@RequestMapping(value = "/api/cars", method = RequestMethod.GET)
	public Page<?> listCars(@RequestParam(value = "page", required = true) int page,
			@RequestParam(value = "limit", required = true) int limit,
			@RequestParam(value = "sortColumn", required = true) String column,
			@RequestParam(value = "sortDir", required = true) String dir) {
		Sort sort = new Sort(
				new Sort.Order("asc".equalsIgnoreCase(dir) ? Sort.Direction.ASC : Sort.Direction.DESC, column));
		Pageable pageRequest = new PageRequest(page, limit, sort);
		Page<Car> p = repo.findAll(pageRequest);
		return p;
	}

	@RequestMapping(value = "/api/cars", method = RequestMethod.PUT)
	public Car updateCar(@RequestBody Car car) {
		Model m = modelRepo.findOne(car.model.id);
		Car c = repo.findOne(car.id);
		if (c != null) {
			c.model = m;
			c.name = car.name;
			c.price = car.price;
			c.year = car.year;
			repo.save(c);
			return c;
		} else {
			return null;
		}
	}

	@RequestMapping(value = "/api/cars", method = RequestMethod.DELETE)
	public Car deleteCar(@RequestBody Car car) {
		repo.delete(car.id);
		return car;
	}

	@RequestMapping(value = "/api/cars", method = RequestMethod.POST)
	public Car createCar(@RequestBody Car car) {
		Model m = modelRepo.findOne(car.model.id);
		Car c = new Car();
		c.name = car.name;
		c.price = car.price;
		c.year = car.year;
		if (m != null) {
			c.model = m;
		}
		repo.save(c);
		return c;
	}

	@RequestMapping(value = "/api/cars/countModelCar", method = RequestMethod.GET)
	public Long countModelCar(@RequestParam(value = "modelId", required = true) String modelId) {
		return repo.countModelCar(new ObjectId(modelId));
	}
}

File: src/main/resources/static/scripts/car_controller.js
app.controller("car_controller", function($scope, $http, $location, $q, ngDialog) {
	
	$scope.page = 0;
	$scope.limit = 5;
	$scope.cars = null;
	$scope.currentPage = 1;
	$scope.totalItems = 0;
	$scope.carId = "";
    $scope.show = true;
    $scope.hide = true;
	$scope.car = {};
	$scope.model = {};
	$scope.models = [];
	$scope.brand = {};
	$scope.brands = [];
	$scope.selectedBrand = null;
	$scope.selectedModel = null;
	$scope.sortColumn = "name";
	$scope.sortDir = "ASC";
	
	$scope.listCars = function() {
		$http.get("/api/cars", {params: {page: $scope.page, limit: $scope.limit, sortColumn: $scope.sortColumn, sortDir: $scope.sortDir}}).then(function(response) {
			$scope.cars = response.data;
			$scope.currentPage = $scope.cars.number+1;
			$scope.totalItems = $scope.cars.totalElements;
		}, function(error) {
			$scope.error = error;
		});
	};
	
	$scope.createCar = function() {
		if(!$scope.validateForm()) {
			$http.post("/api/cars", $scope.car).then(function(response) {
		        $scope.show = true;
		        $scope.hide = true;
		        $scope.hideObj = false;
		        $scope.showObj = false;
		        $scope.carId = "";
		    	$location.path("/car");
			}, function(error) {
				$scope.error = error;
			});
		}
	};
	
	$scope.editCar = function(car) {
        $scope.show = false;
        $scope.hide = false;
        $scope.hideObj = true;
        $scope.showObj = true;
        $scope.carId = car.id;
        $scope.selectedModel = car.model;
        $scope.selectedBrand = car.model.brand;
	};
	
	$scope.deleteCar = function(car) {
		$scope.car = car;
		ngDialog.openConfirm({
			scope: $scope,
			template: 'views/carDeleteConfirm.html',
			className: 'ngdialog-theme-default',
			showClose: false
		}).then(function(value) {
			$http.delete("/api/cars", {data: car, headers: {'Content-Type':'application/json; charset=UTF-8'}}).then(function(response) {
				$scope.listCars();
			}, function(error) {
				$scope.error = error;
			});
		}, function(value) {
		});
	};
	
	$scope.updateCar = function(car) {
		car.model = $scope.selectedModel;
		$http.put("/api/cars", car).then(function(response) {
	        $scope.show = true;
	        $scope.hide = true;
	        $scope.hideObj = false;
	        $scope.showObj = false;
	        $scope.carId = "";
	    	$location.path("/car");
		}, function(error) {
			$scope.error = error;
		});
	};
	
	$scope.undoEdit = function() {
        $scope.show = true;
        $scope.hide = true;
        $scope.hideObj = false;
        $scope.showObj = false;
        $scope.carId = "";
	};
	
	$scope.setPage = function() {
	    $scope.currentPage = pageNo;
	};
	
	$scope.pageChanged = function() {
	    $scope.page = $scope.currentPage - 1;
	    $scope.listCars();
	};
	
	$scope.linkTo = function(path) {
		$location.path(path);
	};
	$scope.listBrands = function() {
		$http.get("/api/brands/listAll").then(function(response) {
			$scope.brands = response.data;
		}, function(error) {
			$scope.error = error;
		});
	};
	
	$scope.listBrandModels = function(brand) {
		$http.get("/api/models/listBrandModels", {params:{brandId: brand.id}}).then(function(response) {
			$scope.models = response.data;
		}, function(error) {
			$scope.error = error;
		});
	};

	$scope.selectedBrandChanged = function() {
		if($scope.car.model == null || $scope.car.model.brand == null || $scope.car.model.brand.id != $scope.selectedBrand.id) {
			$q.all([$scope.listBrandModels($scope.selectedBrand)])
			.then(function(response) {
				$scope.models = response.data;
				$scope.car.brand = $scope.selectedBrand;
				$scope.selectedModel = null;
				delete $scope.car.model;
			});
		}
	};
	
	$scope.selectedModelChanged = function() {
		$scope.car.model = $scope.selectedModel;
	};
	
	$scope.sort = function(column) {
		if($scope.sortColumn == column) {
			$scope.sortDir = $scope.sortDir == "ASC" ? "DESC" : "ASC";
		} else {
			$scope.sortDir = "ASC";
			$scope.sortColumn = column;
		}
		$scope.listCars();
	};
	
	$scope.validateForm = function() {
		$scope.validationMessages = [];
		if($scope.car.name == null || $scope.car.name == null) {
			$scope.validationMessages.push("Name is required.");
		}
		if($scope.car.model == null) {
			$scope.validationMessages.push("Model is not selected.");
		}
		if($scope.car.price == null || $scope.car.price <= 0) {
			$scope.validationMessages.push("Price must be greater thant 0.");
		}
		if($scope.car.year == null || $scope.car.year <= 1962) {
			$scope.validationMessages.push("Year must be greater than 1962.");
		}
		$scope.hasErrors =  $scope.validationMessages.length > 0;
		return $scope.hasErrors;
	};

	
	$q.all([
	    $scope.listBrands(), $scope.listCars()
	]).then(function(response) {
		
	});
});